Userland/Libraries/LibJS/Runtime/RegExpPrototype.cpp

```diff
@@ -87,7 +88,22 @@ static String escape_regexp_pattern(const RegExpObject& regexp_object)
     return pattern;
 }
 
-static void increment_last_index(GlobalObject& global_object, Object& regexp_object)
+// 22.2.5.2.3 AdvanceStringIndex ( S, index, unicode ), https://tc39.es/ecma262/#sec-advancestringindex
+size_t advance_string_index(String const& string, size_t index, bool unicode)
```

Comment:
`size_t` usage here isnt great, since this is supposed to support up to `2**53 - 1`, which is only true for size_t on 64 bit systems, so if we can replace this with a u64 that would be nice (but if its too complicated due to the underlying LibRegex implementation that can be done in a future :)

Reply:
We can probably replace it with `u64`, but the data that ends up as the `index` here is also a `size_t` (comes from `Value::to_length()`), and the `Utf8View` stuff it gets passed to is also dealing with `size_t`

Reply:
yeah, the size_t issue is too prevalent in LibJS already, i would personally still change it here to just not add to the issue (even if it doesnt actually fix the issue on 32 bit yet)

Reply:
I find it extremely unlikely that anyone is gonna have a single string of length >= 2^32 (4GiB), be able to load its entirety _at once_ in any application, and then run a regex match/replace/anything on it - certainly you _could_ make a synthetic test case, but I don't think that'd ever happen in practice :P

Maybe a few years down the line where terabytes of RAM is commonplace?

Reply:
Im not saying that this is a common use case, im just saying were supposed to support it based on the spec, and other engines do as well 🤷

Reply:
I'll take a look at whether they do when I get home (should be simple enough, make a 4GiB string - if you can, then run a regex on it), but I think I remember that Firefox limits strings to 1GiB though.

Reply:
Note that this does not affect only strings, we have the same issue with Arrays, which we also use size_ts for, and i imagine you could more easily find examples of people actually using those at that size, since array storage is sparse.

Reply:
Interesting, V8 also only allows (slightly less than) 1 GiB:
```js
> a = new Array(2 ** 30).join(' '); console.log('done');
Uncaught RangeError: Invalid string length
    at Array.join (<anonymous>)
> a = new Array(2 ** 30 - 3).join(' '); console.log('done');
Uncaught RangeError: Invalid string length
    at Array.join (<anonymous>)
> a = new Array(2 ** 30 - 100).join(' '); console.log('done');
done
undefined
```

(those `console.log`s are just so the REPL wouldn't print `a` after creating it...)

But even with a string length that it accepted, it ran out of heap space creating it:
```
<--- Last few GCs --->

[9444:0x5617485c4e60]    69059 ms: Mark-sweep 2050.5 (2059.7) -> 2050.5 (2052.7) MB, 4.0 / 0.0 ms  (average mu = 0.999, current mu = 0.005) last resort GC in old space requested
[9444:0x5617485c4e60]    69063 ms: Mark-sweep 2050.5 (2052.7) -> 2050.3 (2052.7) MB, 4.3 / 0.0 ms  (average mu = 0.999, current mu = 0.004) last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x29ab52a195d9 <JSObject>
    0: builtin exit frame: parse(aka parse)(this=0x3adff71804b1 <undefined>,0x0727b9000119 <Very long string[1073741780]>,0x3adff71804b1 <undefined>)

    1: 0x026b225cd6e9 <Symbol: onMessage> [0x29ab52a2c5f9] [inspector.js:72] [bytecode=0x168da2a7f581 offset=7](this=0x29ab52a2c881 <EventEmitter map = 0xf77aa7e5059>,0x0727b9000119 <Very long string[1073741780]>)
    2: /* anonymous */ [0x26b225d...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: 0x7f0d88d64a9c node::Abort() [/lib/x86_64-linux-gnu/libnode.so.72]
 2: 0x7f0d88c98872  [/lib/x86_64-linux-gnu/libnode.so.72]
 3: 0x7f0d88f3eaea v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/lib/x86_64-linux-gnu/libnode.so.72]
 4: 0x7f0d88f3edcb v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/lib/x86_64-linux-gnu/libnode.so.72]
 5: 0x7f0d890ec9b9  [/lib/x86_64-linux-gnu/libnode.so.72]
 6: 0x7f0d890ffe84 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/lib/x86_64-linux-gnu/libnode.so.72]
 7: 0x7f0d890c44c1 v8::internal::Factory::AllocateRawWithImmortalMap(int, v8::internal::AllocationType, v8::internal::Map, v8::internal::AllocationAlignment) [/lib/x86_64-linux-gnu/libnode.so.72]
 8: 0x7f0d890cce57 v8::internal::Factory::NewRawOneByteString(int, v8::internal::AllocationType) [/lib/x86_64-linux-gnu/libnode.so.72]
 9: 0x7f0d891e9e3d v8::internal::JsonParser<unsigned char>::MakeString(v8::internal::JsonString const&, v8::internal::Handle<v8::internal::String>) [/lib/x86_64-linux-gnu/libnode.so.72]
10: 0x7f0d891ec3e9 v8::internal::JsonParser<unsigned char>::ParseJsonValue() [/lib/x86_64-linux-gnu/libnode.so.72]
11: 0x7f0d891ed083 v8::internal::JsonParser<unsigned char>::ParseJson() [/lib/x86_64-linux-gnu/libnode.so.72]
12: 0x7f0d88fe7d52  [/lib/x86_64-linux-gnu/libnode.so.72]
13: 0x7f0d88fe835d v8::internal::Builtin_JsonParse(int, unsigned long*, v8::internal::Isolate*) [/lib/x86_64-linux-gnu/libnode.so.72]
14: 0x7f0d89a18779  [/lib/x86_64-linux-gnu/libnode.so.72]
[1]    9444 IOT instruction (core dumped)  js
```

Reply:
(I did update `advance_string_index` to use u64 though to follow the spec)

Reply:
Huh, if they artificially limit String size to 1GB (probably for perf reasons) we could do the same and use size_ts for Strings i guess, not everywhere though.

Reply:
In the real world, unless we split the string to multiple chunks in the engine transparently, a 32-bit system simply _cannot_ go over a size_t, and a 64-bit system is already correct in using a size_t.
I think you guys are making issues out of nothing, unless you can demonstrate that \<thing\> could be done on a 32-bit system without architectural tricks (like the transparent string chunking I mentioned above).

Reply:
(note that I'm not arguing against making array indices correct, that needs to a 64-bit value since it's doable and required)

Reply:
Ali its a fact that the same method he uses here (to_length) is supposed to support all `Array` sizes, and those can be up to 2**53-1, and because the specification uses the same method for Strings, that value is legal. So either we make that value illegal explicitly like the V8 size limit, or we support it, "this doesn't sound useful to me" isnt a valid solution.

Reply:
I'm really only speaking for Regex here, what the other parts of the engine assume or enforce is up to you guys.
If absolutely required, go for it; I'm only saying that it's
- not possible without non-trivial tricks on 32-bit systems
- not particularly useful.

Either way, if it's required, I can help with the regex side of the API.

Reply:
The specification doesnt limit String size artifically, but it doesnt limit TypedArray size either, and we do, and as you said V8 limits String sizes as well, so i think size limiting Strings sounds sensible. (That should let us work with size_t around Strings without worrying about random overflows), if thats what we decide to go with, youre welcome to revert it to size_ts flynn, and then me or someone else can open a PR to add the size check to the String construction methods.

Reply:
:+1: I've put it back to size_t to avoid the 32->64->32 conversions, we can make a max string size explicit in another PR if we decide to do that.